### PR TITLE
update based on text convos

### PR DESCRIPTION
**Pull request by cjlicjli**
definitely still struggling - will continue to look into this tomorrow after work 

```
2023-11-17 01:15:37,281 | INFO | orig sentence: ⲙ̅ⲡϥ̅ⲟⲩⲱϣⲉϭⲱ̅ϣⲁⲁⲧⲉⲡⲣⲟⲑⲉⲥⲙⲓⲁⲙ̅ⲡⲉϥⲁϩⲉ·
2023-11-17 01:15:37,281 | INFO | masked sentence: ⲙ̅ⲡϥ̅ⲟⲩⲱϣ<mask>ϭⲱ̅ϣⲁⲁⲧⲉⲡⲣⲟ<mask>ⲉⲥⲙⲓⲁⲙ̅ⲡⲉϥⲁϩⲉ·

gives this output after each epoch

2023-11-17 01:15:37,302 | INFO |  generated output: ⲉⲉⲉⲉⲉⲉⲙⲙⲙⲙⲙⲙⲙⲙⲙⲙⲙⲙⲙⲙⲙⲉⲙⲙⲙⲙⲙⲙⲙⲙⲙⲙⲙⲉ·
2023-11-17 01:16:33,747 | INFO |  generated output: ⲉⲉⲉ<mask>ⲉⲉⲉⲉⲉⲉⲉⲉⲉⲉⲉⲉⲉⲉⲉ<mask>ⲉⲉⲉⲉⲉⲉⲉ<mask>ⲉⲉ<mask>ⲉⲉⲉ·
2023-11-17 01:17:32,165 | INFO |  generated output: <mask><mask><mask>ⲛ<mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask>ⲛⲛⲛ<mask><mask><mask><mask><mask><mask>·
2023-11-17 01:18:33,791 | INFO |  generated output: <mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask><mask>·
```

### CODE REVIEW

File: coptic_char_generator.py [hunk 1]
```diff
@@ -234,18 +234,38 @@ def train_model(model, train_data, dev_data=None, output_name="charLM"):
...
         logging.info(f" genertated output: {sample}")
 
         torch.save(model, f"{model_path}/{output_name}.pth")
 
     return model
 
 
-def generate(model, seed="The ", n=150, temp=0):
-    # TODO update this to find <mask> tokens and fill them in instead of generating text out
+def fill_masks(model, text, temp=0):
```
Comment (cjlicjli) [inline coptic_char_generator.py:165]: turned your code from earlier into this function instead, just took out the labels b/c they aren't in lookup_indexes anymore

File: coptic_RNN.py [hunk 2]
```diff
@@ -105,7 +106,7 @@ def mask_and_label_characters(self, data_item):
                     replacement = self.mask
                 elif r2 < 0.9:
                     # replace with random character
-                    replacement = random.randint(100, self.num_tokens - 1)
+                    replacement = random.randint(3, self.num_tokens - 1)
```
Comment (cjlicjli) [inline coptic_RNN.py:115]: should start with 0 so the 4 special ones we have are 0-3 (checked with a print in the sentence piece file in case we make any updates)

File: coptic_char_generator.py [hunk 3]
```diff
@@ -147,8 +61,12 @@ def train_batch(model, optimizer, criterion, data, data_indexes, update=True):
...
         index_tensor = torch.tensor(data_item.indexes, dtype=torch.int64).to(device)
         label_tensor = torch.tensor(data_item.labels, dtype=torch.int64).to(device)
-        out, hidden = model([index_tensor])  # [:-1]
-        loss = criterion(out[0], label_tensor)  # [1:]
+        print(label_tensor)
+        out = model([index_tensor])  # [:-1]
+        print(f"out: {out.view(-1, out.shape[-1]).shape}")
+        print(f"out[0]: {out.view(-1, out.shape[-1]).shape}")
+        print(f"label: {label_tensor.view(-1).shape}")
+        loss = criterion(out.view(-1, out.shape[-1]), label_tensor.view(-1))  # [1:]
```
Comment (cjlicjli) [inline coptic_char_generator.py]: spent a lot of time trying different things but haven't made much progress. gonna pick up on looking at the loss section shortly, maybe switch it over to accuracy and see. labels are supposed to make it so that we're only evaluating the masked tokens, but i'm not really understanding how that is happening

**Comment by cjlicjli** [comment]
there's something weird with the loss - getting a bunch of `nan` outputs so will look into that tomorrow and maybe make some of our updates (incorporate accuracy, allow for smart versus random masking) 

did manage to actually fill in the masks correctly when using the full data set (!) - only on the first epoch (the rest of the runs look like the output for the second run)

```Files found: 1403
File read: 41512 sentences
Masking type: random
Training a sentencepiece model
Train coptic_sp model specs: embed_size: 100, hidden_size: 200, proj_size: 150, rnn n layers: 4, share: False, dropout: 0.0
RNN(
  (embed): Embedding(91, 100)
  (scale_up): Linear(in_features=100, out_features=200, bias=True)
  (rnn): LSTM(200, 100, num_layers=4, batch_first=True, bidirectional=True)
  (out): Linear(in_features=200, out_features=100, bias=True)
  (dropout): Dropout(p=0.0, inplace=False)
)
total parameter count = 1,009,100
File coptic_sentences.csv read in with 10001 lines
train=9,501 712,646 694,205 1.0 dev=500 37,862 36,879 1.0 bs=5 lr=0.0003 [100, 200, 150, 4, False, 0.0, 0.15, 91]
1700373103.9204218
0 tr loss      nan      nan -- dev loss      nan      nan -- incremental_batch_size:    5 time elapsed:  138.1
orig sentence: ⲙ̅ⲡϥ̅ⲟⲩⲱϣⲉϭⲱ̅ϣⲁⲁⲧⲉⲡⲣⲟⲑⲉⲥⲙⲓⲁⲙ̅ⲡⲉϥⲁϩⲉ·
masked sentence: ⲙ̅ⲡϥ̅ⲟⲩⲱϣ<mask>ϭⲱ̅ϣⲁⲁⲧⲉⲡⲣⲟ<mask>ⲉⲥⲙⲓⲁⲙ̅ⲡⲉϥⲁϩⲉ·
generated output: ⲙ̅ⲡϥ̅ⲟⲩⲱϣⲉϭⲱ̅ϣⲁⲁⲧⲉⲡⲣⲟⲉⲉⲥⲙⲓⲁⲙ̅ⲡⲉϥⲁϩⲉ·
1700373257.314549
1 tr loss      nan      nan -- dev loss      nan      nan -- incremental_batch_size:    7 time elapsed:  291.5
orig sentence: ⲙ̅ⲡϥ̅ⲟⲩⲱϣⲉϭⲱ̅ϣⲁⲁⲧⲉⲡⲣⲟⲑⲉⲥⲙⲓⲁⲙ̅ⲡⲉϥⲁϩⲉ·
masked sentence: ⲙ̅ⲡϥ̅ⲟⲩⲱϣ<mask>ϭⲱ̅ϣⲁⲁⲧⲉⲡⲣⲟ<mask>ⲉⲥⲙⲓⲁⲙ̅ⲡⲉϥⲁϩⲉ·
generated output: ⲙ̅ⲡϥ̅ⲟⲩⲱϣ<mask>ϭⲱ̅ϣⲁⲁⲧⲉⲡⲣⲟ<mask>ⲉⲥⲙⲓⲁⲙ̅ⲡⲉϥⲁϩⲉ·
1700373422.1892102
```
